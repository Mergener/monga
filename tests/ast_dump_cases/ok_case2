var a : int;

type Vec2 = {
	float : x;
	float : y;
};

function f(a : int, b : int, c : int, d : int, r : res): float {
	var x : int;

	while (a < 5 || b > 6 && c > 12) {
		a = ((c + 3) > 2.05)
			? 20 * 4 + 33 - 12
			: 15 - 5;
		
		b = b * 80;
		c = c * d * 500;

		r.x[2].b = 550;
		r.y = 3;
	}

	return 4 + 33 * d;
}