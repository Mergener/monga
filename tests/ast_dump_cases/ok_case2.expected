<Module>
	<VariableDeclaration>
		<Name>a</Name>
		<Type>int</Type>
	</VariableDeclaration>
	<TypeDefinition>
		<Name>Vec2</Name>
		<TypeDesc>
			<Record>
				<Fields>
					<Field>
						<Type>x</Type>
						<Name>float</Name>
					</Field>
				</Fields>
			</Record>
		</TypeDesc>
	</TypeDefinition>
	<FunctionDefinition>
		<Name>f</Name>
		<ReturnType>float</ReturnType>
		<Parameters>
			<Parameter>
				<Name>a</Name>
				<Type>int</Type>
			</Parameter>
			<Parameter>
				<Name>b</Name>
				<Type>int</Type>
			</Parameter>
			<Parameter>
				<Name>c</Name>
				<Type>int</Type>
			</Parameter>
			<Parameter>
				<Name>d</Name>
				<Type>int</Type>
			</Parameter>
			<Parameter>
				<Name>r</Name>
				<Type>res</Type>
			</Parameter>
		</Parameters>
		<Block>
			<VariableDeclaration>
				<Name>x</Name>
				<Type>int</Type>
			</VariableDeclaration>
			<While>
				<Condition>
					<Or>
						<Condition>
							<LessThan>
								<Var>a</Var>
								<IntConstant>5</IntConstant>
							</LessThan>
						</Condition>
						<Condition>
							<And>
								<Condition>
									<GreaterThan>
										<Var>b</Var>
										<IntConstant>6</IntConstant>
									</GreaterThan>
								</Condition>
								<Condition>
									<GreaterThan>
										<Var>c</Var>
										<IntConstant>12</IntConstant>
									</GreaterThan>
								</Condition>
							</And>
						</Condition>
					</Or>
				</Condition>
				<Do>
					<Assignment>
						<Var>a</Var>
						<ConditionalExp>
							<Condition>
								<GreaterThan>
									<Addition>
										<OperandA>
											<Var>c</Var>
										</OperandA>
										<OperandB>
											<IntConstant>3</IntConstant>
										</OperandB>
									</Addition>
									<FloatConstant>2.050000</FloatConstant>
								</GreaterThan>
							</Condition>
							<ThenExp>
								<Subtraction>
									<OperandA>
										<Addition>
											<OperandA>
												<Multiplication>
													<OperandA>
														<IntConstant>20</IntConstant>
													</OperandA>
													<OperandB>
														<IntConstant>4</IntConstant>
													</OperandB>
												</Multiplication>
											</OperandA>
											<OperandB>
												<IntConstant>33</IntConstant>
											</OperandB>
										</Addition>
									</OperandA>
									<OperandB>
										<IntConstant>12</IntConstant>
									</OperandB>
								</Subtraction>
							</ThenExp>
							<ElseExp>
								<Subtraction>
									<OperandA>
										<IntConstant>15</IntConstant>
									</OperandA>
									<OperandB>
										<IntConstant>5</IntConstant>
									</OperandB>
								</Subtraction>
							</ElseExp>
						</ConditionalExp>
					</Assignment>
					<Assignment>
						<Var>b</Var>
						<Multiplication>
							<OperandA>
								<Var>b</Var>
							</OperandA>
							<OperandB>
								<IntConstant>80</IntConstant>
							</OperandB>
						</Multiplication>
					</Assignment>
					<Assignment>
						<Var>c</Var>
						<Multiplication>
							<OperandA>
								<Multiplication>
									<OperandA>
										<Var>c</Var>
									</OperandA>
									<OperandB>
										<Var>d</Var>
									</OperandB>
								</Multiplication>
							</OperandA>
							<OperandB>
								<IntConstant>500</IntConstant>
							</OperandB>
						</Multiplication>
					</Assignment>
					<Assignment>
						<Field>
							<AccessedExp>
								<Indexed>
									<IndexedExp>
										<Field>
											<AccessedExp>
												<Var>r</Var>
											</AccessedExp>
											<FieldName>x</FieldName>
										</Field>
									</IndexedExp>
									<IndexExp>
										<IntConstant>2</IntConstant>
									</IndexExp>
								</Indexed>
							</AccessedExp>
							<FieldName>b</FieldName>
						</Field>
						<IntConstant>550</IntConstant>
					</Assignment>
					<Assignment>
						<Field>
							<AccessedExp>
								<Var>r</Var>
							</AccessedExp>
							<FieldName>y</FieldName>
						</Field>
						<IntConstant>3</IntConstant>
					</Assignment>
				</Do>
			</While>
			<Return>
				<Addition>
					<OperandA>
						<IntConstant>4</IntConstant>
					</OperandA>
					<OperandB>
						<Multiplication>
							<OperandA>
								<IntConstant>33</IntConstant>
							</OperandA>
							<OperandB>
								<Var>d</Var>
							</OperandB>
						</Multiplication>
					</OperandB>
				</Addition>
			</Return>
		</Block>
	</FunctionDefinition>
</Module>
